### ISSUE TITLE
Allow optional custom random number generators for reproducable results

### ISSUE THREAD

**Issue by FunkMonkey** [cazala/synaptic#117]
We are currently in a phase of finetuning our network parameters, which is unfortunately hard to do, when the output of the neural network is to a certain degree random and not deterministic. 

Thus I propose the possibility to pass an optional random number generator that can be used instead of `Math.random` for initializing biases, weights and for shuffling. Then people can use pseudo random generators like Mersenne Twister.

What do you think?

**Comment by cazala** [comment]
+1, there are several changes i want to push soon, i could try to squeeze this in, but i don't know when that's gonna happen since lately I don't have much time to spend on synaptic. But if you feel like doing this change by yourself feel free to send a PR!

**Comment by FunkMonkey** [comment]
@cazala Good. I can do it, but we should probably decide first on how we will implement it.

Here are the most important occurrences of `Math.random`
- [Neuron bias](https://github.com/cazala/synaptic/blob/116df545fe400ffbc267bb0b6a1ebd4ee2070786/src/neuron.js#L32)
- [Neuron connection weight](https://github.com/cazala/synaptic/blob/116df545fe400ffbc267bb0b6a1ebd4ee2070786/src/neuron.js#L744)
- Neuron bias and connection weight in [reset](https://github.com/cazala/synaptic/blob/116df545fe400ffbc267bb0b6a1ebd4ee2070786/src/neuron.js#L291)
- [Trainer shuffle](https://github.com/cazala/synaptic/blob/116df545fe400ffbc267bb0b6a1ebd4ee2070786/src/trainer.js#L37) (also other occurences in `Trainer`)
- [Architect Liquid](https://github.com/cazala/synaptic/blob/116df545fe400ffbc267bb0b6a1ebd4ee2070786/src/architect.js#L177)

Considering that the `bias` is set in the `Neuron` constructor, we could do the following:
- pass the RNG to the constructor, eg. in the form of `{ rng: ... }` 
- pass the RNG to the constructor of `Neuron.connection` for the `weight`
- pass the RNG to every function that creates a neuron or a connection, f. ex.
 - `Layer` constructor
 - Architect constructors (`Perceptron`, etc.)
 - `Network.fromJSON`

Especially for `Layer` and the architects the question is: should we add these options as a final parameter (backwards-compatible) or do we give them a signature by merging the currently existing options and the proposed one into a single object  (breaking change)? E.g for `Layer`: 
```js
function Layer( options ) {
  const { size, label, rng } = options;
  // ...
}
```

**Comment by Jabher** [comment]
We can use https://github.com/davidbau/seedrandom for that for now, also it can be used by a third-party. Let's move it to v2 expectations?
